### ISSUE TITLE
[FR] filament feed control (jamming)

### ISSUE THREAD

**Issue by maukcc** [MarlinFirmware/Marlin#15345]
At this moment we only have a filament out detection.

We would like a filament jam detection. We used repetier firmware in the past and the way that handled filament jam was very usefull:
It compared extruder steps with incoming pulses from something (an encoder wheel). So every X steps of the extruder (+/- some %) it would require a pulse from the jam detect input.
When the steps, compared to the jam input signal, where off too much, the print speed would be set to xx%
If that fixed the jamming, go back to original speed.
When that did not fix the jamming it would go on pause, marking the extruder jammed.

Especially the reducing speed would automatically fix the jamming.

**Comment by InsanityAutomation** [comment]
This is already there, see the run out and coder settings underneath the filament run-out sensor section.

**Comment by maukcc** [comment]
What we see now is :
#define FILAMENT_MOTION_SENSOR
this expects a pulse from the input before FILAMENT_RUNOUT_DISTANCE_MM has been reached. if it does not see the pulse, it will set filament out and pause the print.
User input required.

The jam sensor reduces the feedrate in an attempt to automatically get rid of the jam and resume the print. (this actually works very well)
No user input required.

**Comment by InsanityAutomation** [comment]
Needs the sensor setup to be quite a bit more precise and i can see alot of false positives triggering a slowdown but its better than a stop. In some cases (hotend too cold, feedrate to fast) this would help. In the most common one I see (heat creep) this would hurt. Regardless not a bad feature to look at implementing as long as its done properly and the time between slowdown attempt and stop isnt too long, or it will defeat the purpose.

**Comment by maukcc** [comment]
We have tested this functionality extensively with repetier firmware and the false positives are non existing because of the margin of error built in the code. 
So if it expects a pulse every 160 pulses of E. it will only generate a jam when it does not see a pulse after 180 pulses, and pauses the print after 300 pulses or x seconds of no pulses.

**Comment by maukcc** [comment]
Can the I2C_POSITION_ENCODERS code be used for this ?

**Comment by Aterfax** [comment]
I was considering making a separate issue but this one seems to mostly match what I was going to discuss:

Filament jamming / pausing / stalling will result in no filament flowing into the hot end - for whatever reason be it a jam on the reel or a loose hobbed gear.

It stands to reason that the hot end will maintain it's temperature - this should result in a lower duty cycle for the hot end heater compared to correct operation and filament melting.

(i.e. melting phase change and heat losses during correct operation vs just heat losses with stalled filament)

I would expect this difference in duty cycle to be significant.

Would it not be possible for the printer firmware to maintain a running variable (or variables) of duty cycle for a given volumetric flow rate and detect a stall via a change in this running value(s)?

This is likely to be material specific, but I can imagine that the printer could be trained by the user for a given material or the printer itself can train itself during the starting stages of the print.

Depending on the sensitivity of this and longevity of data storage - it may also be possible to detect  the loosening of a hobbed gear and decreasing extrusion / under extrusion from the changing duty cycle values. (Though I seriously doubt this would be possible due to changing materials, different temp sensors, different temp sensor placement etc...)

**Comment by maukcc** [comment]
@Aterfax  that is a very intriguing proposition indeed.

**Comment by AnHardt** [comment]
Interesting concept - but difficult!

There are numerous factors having influence on the hot-end-heaters PWM-factor.

Even if we assume:
Power-Supply-Voltage, MOSFETS, MOSFETs temperature, cable length, Heater cartridge power/temperature,
Heater-block material/shape, Nozzle, material/shape, Heat-break material/shape, Thermal grease,
Thermistor/PT100/Thermoelement, wires, reference resistor and it's temperature,
heater-sock, heat-sink,

to be constant during a print the remaining variables are at least:
the set print temperature,
Environment/Chamber temperature and its constancy in the room we move the hot-end in,
Cold-end-Fan on/off/% changing heat-flow over the heat-break and possibly the airflow around the hot-end,
Part-Cooling-Fan on/off/% changing the airflow around the hot-end,
Bed-Heater on/off/% having influence on Power-Supply-Voltage, Temperature in the Environment/Chamber and its distribution,
Coverage of bed by part,
Chamber-door open/closed, Chamber-Fan on/off/%, Chamber-heater on/off/%,
Movement and position of the hot-end, may change the airflow around the nozzle from laminar/convectional to turbulent/forced air-stream,
Filament material, diameter, and finally motion.

All this factors cause different magnitudes of change in the hot-end-PWM, all of them have nonlinear effects and all of them have different time constants (from switching to the new state to seeing the  effect in the temperature/PWM-ratio).

PID is a simple thing, it just reacts on the summ of effects to the temperature, the difference to target temperature, the duration of the difference and the speed of change.

But reliably isolating the information, if the filament is moving (enough), from the PWM-ratio, seems to be a bit challenging.

A more direct way to measure filament motion seems to be more promising.
We don't run on mainframes with deep learning systems.

**Comment by Aterfax** [comment]
Yeah I figure getting any fine detail out of it is probably a no go for someone without a nvidia GPGPU handy, though I would have thought the duty cycle difference between melting and not melting would probably be significant enough (the amount of energy to melt something is typically 10 to 100 times its heat capacity) and not require any "deep learning."

I'd check on my own machine if I knew how :)

On a simple basis, the firmware would simply be looking for a dramatic change in duty cycle when ruling out other factors (e.g. a lot of travel moves and no extrusion.)

**Comment by maukcc** [comment]
I am guessing that most feature-requests get stale
